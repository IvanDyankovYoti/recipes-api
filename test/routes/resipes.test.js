const request = require('supertest');
const fs = require('fs');
const path = require('path');
const server = require('../../app');

describe('/recipes', () => {
  const dataFile = path.resolve(__dirname, '../../seedData/recipe-data.csv');
  const dataFileBkp = path.resolve(__dirname, '../../seedData/recipe-data_bkp.csv');

  beforeAll(() => {
    // copy the database, becaues the tests will change it
    fs.createReadStream(dataFile)
      .pipe(fs.createWriteStream(dataFileBkp));
  });

  afterAll(() => {
    fs.unlinkSync(dataFile); // delete the modified data
    fs.renameSync(dataFileBkp, dataFile); // restore
    server.close();
  });

  describe('GET /recipes/:id', () => {
    it('should return 200 OK with recipe JSON data', (done) => {
      request(server)
        .get('/recipes/1')
        .expect('Content-Type', /json/)
        .expect(200)
        .end((err, data) => {
          if (err) return done(err);
          expect(data.body).toMatchSnapshot();
          return done();
        });
    });
    it('should return 404 Not Found when recipe is not found', (done) => {
      request(server)
        .get('/recipes/INVALID_INDEX')
        .expect(404, done);
    });
  });

  describe('GET /recipes', () => {
    it('should return 200 OK with 10 available recipes', (done) => {
      request(server)
        .get('/recipes')
        .expect('Content-Type', /json/)
        .expect(200)
        .end((err, data) => {
          if (err) return done(err);
          expect(data.body).toMatchSnapshot();
          return done();
        });
    });

    it('should return 200 OK with recipes for specified recipe_cuisine', (done) => {
      request(server)
        .get('/recipes?recipe_cuisine=mexican')
        .expect('Content-Type', /json/)
        .expect(200)
        .end((err, data) => {
          if (err) return done(err);
          expect(data.body).toMatchSnapshot();
          return done();
        });
    });
  });

  describe('PUT /recipes', () => {
    it('should return 200 OK with the updated recipe', (done) => {
      request(server)
        .put('/recipes/1')
        .send({ description: 'I have changed', title: 'Changed title' })
        .expect('Content-Type', /json/)
        .expect(200)
        .end((err, data) => {
          if (err) return done(err);
          expect(data.body).toMatchSnapshot();
          return done();
        });
    });

    it('should return 404 Not Found when the recipe is not found', (done) => {
      request(server)
        .put('/recipes/INVALID_INDEX')
        .send({ description: 'I have changed', title: 'Changed title' })
        .expect(404, done);
    });

    // Enable me when functionality is implemented
    it.skip('should return 400 Bad request when trying to update invalid fields', (done) => {
      request(server)
        .put('/recipes/INVALID_INDEX')
        .send({ description: 'I have changed', title: 'Changed title', nonExisting: true })
        .expect(400, done);
    });

    // Enable me when functionality is implemented
    it.skip('should return 400 Bad request when trying to update field with invalid data', (done) => {
      request(server)
        .put('/recipes/INVALID_INDEX')
        .send({ description: true, title: 'Changed title' })
        .expect(400, done);
    });
  });
});
