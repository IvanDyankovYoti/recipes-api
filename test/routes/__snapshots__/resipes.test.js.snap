// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/recipes GET /recipes should return 200 OK with 10 available recipes 1`] = `
Array [
  Object {
    "description": "Here we've used onglet steak which is an extra flavoursome cut of beef that should never be cooked past medium rare. So if you're a fan of well done steak, this one may not be for you. However, if you love rare steak and fancy trying a new cut, please be",
    "id": "1",
    "title": "UPDATED Sweet Chilli and Lime Beef on a Crunchy Fresh Noodle Salad",
  },
  Object {
    "description": "Tamil Nadu is a state on the eastern coast of the southern tip of India. Curry from there is particularly famous and it's easy to see why. This one is brimming with exciting contrasting tastes from ingredients like chilli powder, coriander and fennel seed",
    "id": "2",
    "title": "Tamil Nadu Prawn Masala",
  },
  Object {
    "description": "This delicious pasta dish comes from the Italian region of Umbria. It has a smoky and intense wild boar flavour which combines the earthy garlic, leek and onion flavours, while the chilli flakes add a nice deep aroma. Enjoy within 5-6 days of delivery.",
    "id": "3",
    "title": "Umbrian Wild Boar Salami Ragu with Linguine",
  },
  Object {
    "description": "One for those who like their veggies with a slightly spicy kick. However, those short on time, be warned ' this is a time-consuming dish, but if you're willing to spend a few extra minutes in the kitchen, the fresh corn mash is extraordinary and worth a t",
    "id": "4",
    "title": "Tenderstem and Portobello Mushrooms with Corn Polenta",
  },
  Object {
    "description": "A classic roast with a twist. The pork loin is marinated in rosemary, fennel seeds and chilli flakes then teamed with baked potato wedges and butter beans in tomato sauce. Enjoy within 5-6 days of delivery.",
    "id": "5",
    "title": "Fennel Crusted Pork with Italian Butter Beans",
  },
  Object {
    "description": "Succulent pork tenderloin and feathery white bean and parsnip mash mingle with feisty cumin seeds and tangy leek in this lighter, less conventional take on a British classic. Welcome to the outer limits of food!",
    "id": "6",
    "title": "Pork Chilli",
  },
  Object {
    "description": "Kick-start the new year with some get-up and go with this lean green vitality machine. Protein-packed chicken and mineral-rich kale are blended into a smooth, nut-free version of pesto; creating the ultimate composition of nutrition and taste",
    "id": "7",
    "title": "Courgette Pasta Rags",
  },
  Object {
    "description": "A Goustofied British institution, learn how to make beautifully golden breaded chicken escalopes drizzled in homemade garlic butter and served atop fluffy potato and broccoli mash.",
    "id": "8",
    "title": "Homemade Eggs & Beans",
  },
  Object {
    "description": "I love this super healthy fish dish, it contains a punch from zingy ginger, a kick from chili and a salty sweet balance from soy sauce and mirim. A cleansing and restorative meal, great for body and soul.",
    "id": "9",
    "title": "Grilled Jerusalem Fish",
  },
]
`;

exports[`/recipes GET /recipes should return 200 OK with recipes for specified recipe_cuisine 1`] = `
Array [
  Object {
    "description": "Comprising all the best bits of the classic American number and none of the mayo, this is a warm & tasty chicken and bulgur salad with just a hint of Scandi influence. A beautifully summery medley of flavours and textures",
    "id": "10",
    "title": "Pork Katsu Curry",
  },
]
`;

exports[`/recipes GET /recipes/:id should return 200 OK with recipe JSON data 1`] = `
Object {
  "base": "noodles",
  "box_type": "vegetarian",
  "bulletpoint1": "",
  "bulletpoint2": "",
  "bulletpoint3": "",
  "carbs_grams": "0",
  "created_at": "30/06/2015 17:58:00",
  "equipment_needed": "Appetite",
  "fat_grams": "35",
  "gousto_reference": "59",
  "icalories_kcald": "401",
  "id": "1",
  "in_your_box": "",
  "marketing_description": "Here we've used onglet steak which is an extra flavoursome cut of beef that should never be cooked past medium rare. So if you're a fan of well done steak, this one may not be for you. However, if you love rare steak and fancy trying a new cut, please be",
  "origin_country": "Great Britain",
  "preparation_time_minutes": "35",
  "protein_grams": "12",
  "protein_source": "",
  "recipe_cuisine": "asian",
  "recipe_diet_type_id": "meat",
  "season": "all",
  "shelf_life_days": "4",
  "short_title": "",
  "slug": "sweet-chilli-and-lime-beef-on-a-crunchy-fresh-noodle-salad",
  "title": "UPDATED Sweet Chilli and Lime Beef on a Crunchy Fresh Noodle Salad",
  "updated_at": "30/06/2015 17:58:00",
}
`;

exports[`/recipes PUT /recipes should return 200 OK with the updated recipe 1`] = `
Object {
  "base": "noodles",
  "box_type": "vegetarian",
  "bulletpoint1": "",
  "bulletpoint2": "",
  "bulletpoint3": "",
  "carbs_grams": "0",
  "created_at": "30/06/2015 17:58:00",
  "description": "I have changed",
  "equipment_needed": "Appetite",
  "fat_grams": "35",
  "gousto_reference": "59",
  "icalories_kcald": "401",
  "id": "1",
  "in_your_box": "",
  "marketing_description": "Here we've used onglet steak which is an extra flavoursome cut of beef that should never be cooked past medium rare. So if you're a fan of well done steak, this one may not be for you. However, if you love rare steak and fancy trying a new cut, please be",
  "origin_country": "Great Britain",
  "preparation_time_minutes": "35",
  "protein_grams": "12",
  "protein_source": "",
  "recipe_cuisine": "asian",
  "recipe_diet_type_id": "meat",
  "season": "all",
  "shelf_life_days": "4",
  "short_title": "",
  "slug": "sweet-chilli-and-lime-beef-on-a-crunchy-fresh-noodle-salad",
  "title": "Changed title",
  "updated_at": "30/06/2015 17:58:00",
}
`;
